File: app/src/main/java/vinhlong/ditagis/com/khaosatdongho/async/RefreshVatTuAsync.java
package vinhlong.ditagis.com.khaosatdongho.async;

import android.app.Activity;
import android.os.AsyncTask;
import android.support.design.widget.BottomSheetDialog;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.esri.arcgisruntime.concurrent.ListenableFuture;
import com.esri.arcgisruntime.data.Feature;
import com.esri.arcgisruntime.data.FeatureQueryResult;
import com.esri.arcgisruntime.data.QueryParameters;
import com.esri.arcgisruntime.data.ServiceFeatureTable;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ExecutionException;

import vinhlong.ditagis.com.khaosatdongho.R;
import vinhlong.ditagis.com.khaosatdongho.adapter.VatTuApdapter;
import vinhlong.ditagis.com.khaosatdongho.libs.Action;
import vinhlong.ditagis.com.khaosatdongho.utities.Constant;

/**
 * Created by ThanLe on 4/16/2018.
 */

public class RefreshVatTuAsync extends AsyncTask<String, List<VatTuApdapter.VatTu>, Void> {
    private BottomSheetDialog mDialog;
    private Activity mActivity;
    private ServiceFeatureTable vatTuTable;
    private VatTuApdapter vatTuApdapter;
    private Action action;
    private ArrayList<Feature> dmVatTuFeatures;

    public interface AsyncResponse {
        void processFinish(List<Feature> features);
    }

    private AsyncResponse delegate;

    public RefreshVatTuAsync(Activity activity, ServiceFeatureTable vatTuTable, ArrayList<Feature> features, VatTuApdapter vatTuApdapter, Action action, AsyncResponse asyncResponse) {
        this.delegate = asyncResponse;
        mActivity = activity;
        this.vatTuTable = vatTuTable;
        this.vatTuApdapter = vatTuApdapter;
        this.action = action;
        this.dmVatTuFeatures = features;
    }

    @Override
    protected void onPreExecute() {
        super.onPreExecute();
        mDialog = new BottomSheetDialog(this.mActivity);
        LinearLayout view = (LinearLayout) mActivity.getLayoutInflater().inflate(R.layout.layout_progress_dialog, null, false);
        ((TextView) view.findViewById(R.id.txt_progress_dialog_title)).setText("Đang xử lý...");
        mDialog.setContentView(view);
        mDialog.setCancelable(false);

        mDialog.show();

    }

    @Override
    protected Void doInBackground(String... params) {
        final List<Feature> features = new ArrayList<>();
        final List<VatTuApdapter.VatTu> vatTus = new ArrayList<>();
        QueryParameters queryParameters = new QueryParameters();
        String queryClause = Constant.VatTuFields.MaKhachHang + " = '" + params[0] + "'";
        queryParameters.setWhereClause(queryClause);
        final ListenableFuture<FeatureQueryResult> queryResultListenableFuture = vatTuTable.queryFeaturesAsync(queryParameters, ServiceFeatureTable.QueryFeatureFields.LOAD_ALL);
        queryResultListenableFuture.addDoneListener(() -> {
            try {
                FeatureQueryResult result = queryResultListenableFuture.get();
                Iterator iterator = result.iterator();

                while (iterator.hasNext()) {
                    Feature feature = (Feature) iterator.next();
                    VatTuApdapter.VatTu vatTu = new VatTuApdapter.VatTu();
                    Object maVatTu = feature.getAttributes().get(Constant.VatTuFields.MaVatTu);
                    Object soLuong = feature.getAttributes().get(Constant.VatTuFields.SoLuong);
                    Object objectID = feature.getAttributes().get(Constant.LayerFields.OBJECTID);
                    vatTu.setMaVatTu(objectID.toString());
                    if (soLuong != null) {
                        vatTu.setSoLuongVatTu(soLuong.toString());
                    }
                    if (maVatTu != null) {
                        Feature loaiVatTu = getLoaiVatTu(maVatTu.toString());
                        if (loaiVatTu != null) {
                            Object donViTinh = loaiVatTu.getAttributes().get(Constant.LoaiVatTuFields.DonViTinh);
                            Object giaVatTu = loaiVatTu.getAttributes().get(Constant.LoaiVatTuFields.GiaVatTu);
                            Object tenVatTu = loaiVatTu.getAttributes().get(Constant.LoaiVatTuFields.TenVatTu);
                            if (donViTinh != null)
                                vatTu.setDonViTinh(donViTinh.toString());
                            if (giaVatTu != null)
                                vatTu.setGiaNC(giaVatTu.toString());
                            if (tenVatTu != null)
                                vatTu.setTenVatTu(tenVatTu.toString());
                        }
                    }
                    vatTus.add(vatTu);
                    features.add(feature);
                }
                publishProgress(vatTus);
                delegate.processFinish(features);
            } catch (InterruptedException e) {
                e.printStackTrace();
                publishProgress();
            } catch (ExecutionException e) {
                e.printStackTrace();
                publishProgress();
            }
        });
        return null;
    }

    private Feature getLoaiVatTu(String maVatTu) {
        if (maVatTu != null) {
            for (Feature feature : this.dmVatTuFeatures) {
                Map<String, Object> attributes = feature.getAttributes();
                Object maVT = attributes.get(Constant.VatTuFields.MaVatTu);
                if (maVT != null && maVT.toString().equals(maVatTu)) {
                    return feature;
                }
            }
        }
        return null;
    }

    @Override
    protected void onProgressUpdate(List<VatTuApdapter.VatTu>... values) {
        vatTuApdapter.clear();
//        vatTuApdapter.setVatTus(values[0]);
        vatTuApdapter.notifyDataSetChanged();
        if (mDialog != null && mDialog.isShowing()) mDialog.dismiss();
        super.onProgressUpdate(values);

    }


    @Override
    protected void onPostExecute(Void result) {
        super.onPostExecute(result);

    }

}

File: app/src/main/java/vinhlong/ditagis/com/khaosatdongho/async/RefreshVatTuAsync.kt
package vinhlong.ditagis.com.khaosatdongho.async

import android.annotation.SuppressLint
import android.app.Activity
import android.os.AsyncTask
import com.esri.arcgisruntime.data.Feature
import com.esri.arcgisruntime.data.QueryParameters
import com.esri.arcgisruntime.data.ServiceFeatureTable
import vinhlong.ditagis.com.khaosatdongho.adapter.VatTuApdapter
import vinhlong.ditagis.com.khaosatdongho.libs.Action
import vinhlong.ditagis.com.khaosatdongho.utities.Constant
import java.util.*
import java.util.concurrent.ExecutionException

/**
 * Created by ThanLe on 4/16/2018.
 */
@SuppressLint("StaticFieldLeak")
class RefreshVatTuAsync( private val mActivity: Activity, private val vatTuTable: ServiceFeatureTable, private val dmVatTuFeatures: ArrayList<Feature>, private val vatTuApdapter: VatTuApdapter, private val action: Action, private val delegate: AsyncResponse) : AsyncTask<Long, List<VatTuApdapter.VatTu>, Void>() {

    interface AsyncResponse {
        fun processFinish(features: List<Feature>?)
    }

     override fun doInBackground(vararg params: Long?): Void? {
        val features = ArrayList<Feature>()
        val vatTus = ArrayList<VatTuApdapter.VatTu>()
        val queryParameters = QueryParameters()
        val queryClause = Constant.VatTuFields.MaKhachHang + " = '" + params[0] + "'"
        queryParameters.whereClause = queryClause
        val queryResultListenableFuture = vatTuTable.queryFeaturesAsync(queryParameters, ServiceFeatureTable.QueryFeatureFields.LOAD_ALL)
        queryResultListenableFuture.addDoneListener {
            try {
                val result = queryResultListenableFuture.get()
                val iterator = result.iterator()

                while (iterator.hasNext()) {
                    val feature = iterator.next() as Feature
                    val vatTu = VatTuApdapter.VatTu()
                    val maVatTu = feature.attributes[Constant.VatTuFields.MaVatTu]
                    val soLuong = feature.attributes[Constant.VatTuFields.SoLuong]
                    val objectID = feature.attributes[Constant.LayerFields.OBJECTID]
                    vatTu.maVatTu = objectID.toString()
                    if (soLuong != null) {
                        vatTu.setSoLuongVatTu(soLuong as Double)
                    }
                    if (maVatTu != null) {
                        val loaiVatTu = getLoaiVatTu(maVatTu.toString())
                        if (loaiVatTu != null) {
                            val donViTinh = loaiVatTu.attributes[Constant.LoaiVatTuFields.DonViTinh]
                            val giaVatTu = loaiVatTu.attributes[Constant.LoaiVatTuFields.GiaVatTu]
                            val tenVatTu = loaiVatTu.attributes[Constant.LoaiVatTuFields.TenVatTu]
                            if (donViTinh != null)
                                vatTu.donViTinh = donViTinh.toString()
                            if (giaVatTu != null)
                                vatTu.setGiaNC(giaVatTu.toString())
                            if (tenVatTu != null)
                                vatTu.tenVatTu = tenVatTu.toString()
                        }
                    }
                    vatTus.add(vatTu)
                    features.add(feature)
                }
                publishProgress(vatTus)
                delegate.processFinish(features)
            } catch (e: InterruptedException) {
                e.printStackTrace()
                publishProgress()
            } catch (e: ExecutionException) {
                e.printStackTrace()
                publishProgress()
            }
        }
        return null
    }

    private fun getLoaiVatTu(maVatTu: String?): Feature? {
        if (maVatTu != null) {
            for (feature in this.dmVatTuFeatures) {
                val attributes = feature.attributes
                val maVT = attributes[Constant.VatTuFields.MaVatTu]
                if (maVT != null && maVT.toString() == maVatTu) {
                    return feature
                }
            }
        }
        return null
    }

    override fun onProgressUpdate(vararg values: List<VatTuApdapter.VatTu>) {
        vatTuApdapter.clear()
        //        vatTuApdapter.setVatTus(values[0]);
        vatTuApdapter.notifyDataSetChanged()
        super.onProgressUpdate(*values)

    }

}

